deku-derive/src/macros/deku_read.rs

```diff
@@ -348,20 +348,31 @@ fn emit_field_read(
     } else {
         let read_args = gen_field_args(field_endian, f.bits, f.ctx.as_ref())?;
 
-        // Count is special, we need to generate `(count, (other, ..))` for it.
+        // The container limiting options are special, we need to generate `(limit, (other, ..))` for them.
+        // These have a problem where when it isn't a copy type, the field will be moved.
+        // e.g. struct FooBar {
+        //   a: Baz // a type implement `Into<usize>` but not `Copy`.
+        //   #[deku(count = "a") <-- Oops, use of moved value: `a`
+        //   b: Vec<_>
+        // }
         if let Some(field_count) = &f.count {
-            // The count has same problem, when it isn't a copy type, the field will be moved.
-            // e.g. struct FooBar {
-            //   a: Baz // a type implement `Into<usize>` but not `Copy`.
-            //   #[deku(count = "a") <-- Oops, use of moved value: `a`
-            //   b: Vec<_>
-            // }
             quote! {
                 {
                     use core::borrow::Borrow;
                     DekuRead::read(rest, (usize::try_from(*((#field_count).borrow()))?.into(), (#read_args)))
                 }
             }
+        } else if let Some(field_bits) = &f.bits_read {
+            quote! {
+                {
+                    use core::borrow::Borrow;
+                    DekuRead::read(rest, (deku::ctx::BitSize(usize::try_from(*((#field_bits).borrow()))?).into(), (#read_args)))
+                }
+            }
+        } else if let Some(field_until) = &f.until {
+            // We wrap the input into another closure here to enforce that it is actually a callable
+            // Otherwise, an incorrectly passed-in integer could unexpectedly convert into a `Count` limit
```

Comment:
We could remove the `From/Into` impl on `Limit` and enforce an explicit `Limit` be initialized? This may make it more readable as well.

Reply:
The problem with that is that I _think_ it means we need to specify the `Limit` type parameters here, which is non-trivial to do?

I've just now thought that it might be nicer to make the attribute take an expression instead of a callable directly, and then this line becomes:
```rust
             quote! {DekuRead::read(rest, ((|value: &_| #field_until).into(), (#read_args)))} 
```

This feels much nicer, and also means the user doesn't need to explicity mark the type in their lambda, as they just have to specify an expression using `value`?

Reply:
Specifying a lambda makes it explicit from the user and no need to remember/know that `value` is available/exists. Comes back to the `internal available variables` discussion in https://github.com/sharksforarms/deku/issues/111

I prefer the long form in this case, if we can avoid having "internal" variables

Reply:
A possible alternative could be to write functions on the `Limit` enum and use them to enforce type arguments

```rust
impl Limit {
fn new_count(v: usize) {...}
fn new_bits(v: BitSize) {...}
fn new_until(v: FnMut) {...}
}
```

Reply:
> Specifying a lambda makes it explicit from the user and no need to remember/know that value is available/exists. 

Yeah, good point, it's much nicer to be explicit

> A possible alternative could be to write functions on the Limit enum and use them to enforce type arguments

For some reason I thought this wouldn't nicely work, but then I remembered it obviously does because `Vec::new` doesn't require explicit arguments to `Vec` 🤦